client/cmd/dexcctl/main.go

```diff
@@ -41,6 +43,62 @@ func main() {
 	}
 }
 
+// promptPasswords is a map of routes to password prompts. Passwords are
+// prompted in the order given.
+var promptPasswords = map[string][]string{
+	"openwallet": {"App password:"},
+	"newwallet":  {"App password:", "Wallet password:"},
+	"init":       {"Set new app password:"},
+	"register":   {"App password:"},
+}
+
+// readCerts is a map of routes to TLS certificate indexes.
+var readCerts = map[string]int{
+	"preregister": 1,
+	"register":    2,
+}
+
+// promptPWs prompts for passwords on stdin and returns an error if prompting
+// fails or a password is empty. Returns passwords as a slice of strings.
+func promptPWs(cmd string) ([]string, error) {
+	prompts, exists := promptPasswords[cmd]
+	if !exists {
+		return nil, nil
+	}
+	pws := make([]string, len(prompts))
+	// Prompt for passwords one at a time.
+	for i, prompt := range prompts {
+		pw, err := admin.PasswordPrompt(prompt)
+		if err != nil {
+			return nil, err
+		}
+		pws[i] = string(pw)
+		admin.ClearBytes(pw)
+	}
+	return pws, nil
+}
+
+// readCert reads TLS certificate content located in a file specified at args'
+// index as expected for cmd, replaces it with the file as a string, and returns
+// the altered params.
+func readCert(cmd string, args []string) (params []string, err error) {
+	idx, exists := readCerts[cmd]
+	// Certs are optional, so it is not an error if they are not included.
+	if !exists || len(args) < idx+1 || args[idx] == "" {
+		return args, nil
+	}
+	path := cleanAndExpandPath(args[idx])
+	if !fileExists(path) {
+		return nil, fmt.Errorf("no cert file found at %s", path)
+	}
+	certB, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("error reading %s: %v", path, err)
+	}
+	params = append(append(args[:idx], string(certB)), args[idx+1:]...)
```

Comment:
You can modify the slice directly within a function, you just can't modify the length. `params[idx] = string(certB)` is fine. 

Or maybe it's better to have `readCert` return  the certificate `string`, and have the caller do the replacement. 

Any preference, @chappjc?

Reply:
Hmm, i missed this comment.  I only saw https://github.com/decred/dcrdex/pull/246/commits/ff84c554b4db1bb37f5ed92bf156748c4d8afdaa.  :/

I was fine with either a string return for the cert or replacing the filename with the contents.  Does the caller need both the path and the contents?  We've gone back and forth and I don't follow the current commit.

Reply:
But regarding modifying the input string, `args` is the input and that does not seem to be modified.  Still, I think this is almost as confusing as the original version with the `*string`.  After reading the cert, does the caller still need the cert path name?

Reply:
We can talk it through on matrix when we're all online.  But I preferred a function that just identifies the cert path arg, loads it, and returns the cert contents.  If the cert path is no longer required, the function could replace rather than add to the params/args slice.  Or the caller can do the replace.  That choice isn't such a big deal to me; it's more an issue of making the semantics of the function clear and making it less error prone.

Reply:
I might still be missing something, but I think the easiest solution right now is to just 

```
func readCert(cmd string, args []string) error {
    ...
    args[idx] = string(certB)
    return nil
}
```

Reply:
If it's preferable to have the cert contents itself in the slice passed to core, yeah, I agree that unless we still need the cert path for some reason then that seems to be the most straightforward solution.

Reply:
Gotcha. `readCert` would just go away altogether then. `loadCert` in your previous example would just be `ioutil.ReadFile`. 

That is cleaner.